File: leetcode/editor/en/[2001]Number of Pairs of Interchangeable Rectangles.js
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * @param {number[][]} rectangles
 * @return {number}
 */
var interchangeableRectangles = function (rectangles) {
    let map = new Map();
    let count = 0;

    for (const [x, y] of rectangles) {
        const ratio = x / y;
        if (!map.has(ratio)) {
            map.set(ratio, 1);
        } else {
            map.set(ratio, map.get(ratio) + 1);
        }
    }

    for (const num of map.values()) {
        if (num > 1) {
            count += num * (num - 1) / 2;
        }
    }

    return count;
};
//leetcode submit region end(Prohibit modification and deletion)

File: leetcode/editor/en/[2001]Number of Pairs of Interchangeable Rectangles.java

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public long interchangeableRectangles(int[][] rectangles) {
        Map<Double, Long> map = new HashMap<>();
        long count = 0;

        for (int[] rec : rectangles) {
            int x = rec[0];
            int y = rec[1];
            double ratio = (double) x / y;

            map.put(ratio, map.getOrDefault(ratio, 0L) + 1);
        }

        for (long num : map.values()) {
            if (num > 1) {
                count += num * (num - 1) / 2;
            }
        }

        return count;
    }
}
//leetcode submit region end(Prohibit modification and deletion)
